chore(vote): stubbed out voting

nguernse opened the pull request:
# Pull Request Template

- Fixes: <!-- #(issue) -->
- Closes: <!-- #(issue) -->

## Checklist

- [ ] This PR conforms to the guidelines in [CONTRIBUTING.md](../CONTRIBUTING.md)
  - [ ] It has a descriptive title
  - [ ] My commits conform to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification
- [ ] This PR is not a duplicate of a similar PR ([how to search issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) for similar/duplicate PRs, which this is not
- [ ] This PR targets the `dev` branch
- [ ] I ran the pre-commit hooks or `npm test && npm lint && npm test:e2e` and all tests pass

<hr>

## Description

<!--
Please include a summary of the change and which issue it addresses. Include the motivation for the changes and any additional context.
-->

## Type of Change

<!--
Please delete options that are not relevant.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code cleanliness (removal of unnecessary comments or debug logs)

## How Has This Been Tested?

<!--
Please describe the tests that you ran to verify your changes.
- Please Provide instructions so we can reproduce the scenarios.
- Please List any relevant details for your test configuration.

    - [ ] Test A
    - [ ] Test B
-->
<!--
**Test Configuration** (Optional):

- Firmware version:
    -  If you know the version of any specific software you used, list it here. If unsure, you can leave it blank.
- Hardware:
    -  Mention your computer or device model if your changes are hardware-dependent. If not, it's okay to skip.
- Toolchain:
    - This refers to the set of programming tools you used. For beginners, it could be the IDE (Integrated Development Environment) like Visual Studio Code.
- SDK:
    - Mention any specific Software Development Kit versions you used. If you didn't use any, no need to include it.
-->

## Microservices and API Integration

- [ ] If applicable, my changes are compatible with the project's microservices architecture and API gateway integration requirements.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have made corresponding changes to the documentation
- [ ] I have removed all unnecessary comments and debug log statements
- [ ] I have linked this PR to its corresponding issue

## Screenshots (if appropriate):

Please review and suggest any further insights, improvements, or modifications needed.

helmturner reviewed (changes requested):
ISSUE(blocking): This needs updating in light of the changes in #236. 
SUGGESTION: Might be easiest to split each component into it's own file, and then rebase on top of #236 (or main, if it's merged). That way you have fewer merge conflicts and can compare before/after a little easier.
NOTE: Happy to do this for you, if you prefer!

helmturner commented:
I took the liberty of handling the merge conflicts and bringing this up to date with main, but I didn't want to force push over your pr branch without giving you a chance to review it first.

There's a copy of the results at https://github.com/tulsawebdevs/website/tree/feature/proposal-voting in case you'd like to compare them. If that branch looks good to you, let me know and I'll push it to this PR.

nguernse commented:
@helmturner Awesome, thanks for doing that! I had a look and everything looks good to me.

nguernse commented:
@helmturner I had some time today, so hope you don't mind I went ahead and incorporated your updates into this PR. I then also added
- debouncing voting
- preventing votes without being authenticated
- a few UI updates to proposal card

Regarding the voting data: Will the user's voting data (upvotes and totals depending on proposal status) be included with the proposal data or will we be needing to fetch that on a per-proposal basis?

helmturner commented:
Oh nice, thank you! I'm sorry I didn't get around to it sooner, some family stuff came up that I had to focus on. 

Re: votes, I think the idea will be to have an endpoint specifically for user's votes. That would be 1 query for user's voting record (maybe we pull the whole thing and put it in client state), and then when a proposal/proposal list is queried we can just populate the vote totals from state.

I'll get this reviewd right now 🖖

helmturner commented on src/features/vote/proposals/list/ProposalListItem.tsx:
```diff
@@ -40,17 +47,66 @@ export default function ProposalListItem({ proposal }: Props) {
+    `https://vote.tulsawebdevs.org/proposals/${proposal.id}/vote`,
+  );
+
+  const handleVote = useProtectedFunction(
+    useDebounce((value: Vote) => {
+      const votePayload: VotePayload = {
+        initiativeId: proposal.id.toString(10),
+        vote: value,
+        comment: '',
+        authorId: proposal.authorId,
+        authorName: proposal.authorName,
+        authorEmail: proposal.authorEmail,
+      };
+
+      setVoteValue(value);
+      voteRequest(votePayload).catch(() => {
+        toast.error('Could not cast vote. Please try again.');
+      });
+    }),
+    {
+      unauthorizedMessage: 'You do not have permission to vote.',
+    },
+  );
```
ISSUE: We probably want to take the `setVoteValue(value)` call out of the debounce to improve perceived performance. As it is, some clicks are 'ignored' from the user's perspective.
FIX: If I'm not mistaken, we should just be able to lift the function call out of the debounce hook without causing the value to fall out of sync with what's sent in the request.
NOTE(non-blocking): Normally we'd want to store the user's initial vote in memory and only update it when we get a success confirmation from the server. That way we can roll back the UI if the vote request fails. No need to worry about it here, though; i'm researching some libraries to use for data fetching that should abstract most of the query management cycle away.

nguernse replied:
That makes sense. I thought it was awkward to see a slight delay. I've lifted up setting the vote value.

helmturner commented on src/features/vote/proposals/list/ProposalInterestVote.tsx:
```diff
@@ -0,0 +1,95 @@
+import { cn } from '../../../../utils.ts';
+import type { Proposal, Vote } from '../types.ts';
+
+const voteOptions = [
+  {
+    label: 'Strongly Disinterested',
+    value: '-2',
+    id: 'vote-strongly-disinterested',
+  },
+  {
+    label: 'Slightly Disinterested',
+    value: '-1',
+    id: 'vote-slightly-disinterested',
+  },
+  { label: 'Neutral', value: '0', id: 'vote-neutral' },
+  { label: 'Slightly Interested', value: '1', id: 'vote-slightly-interested' },
+  { label: 'Strongly Interested', value: '2', id: 'vote-strongly-interested' },
+];
```
POLISH: I think we can shorten some of the labels on these
SUGGESTION: 
```suggestion
const voteOptions = [
  {
    label: 'Strong Disinterest',
    value: '-2',
    id: 'vote-strong-disinterest',
  },
  {
    label: 'Slight Disinterest',
    value: '-1',
    id: 'vote-slight-disinterest',
  },
  { label: 'Neutral', value: '0', id: 'vote-neutral' },
  { label: 'Slight Interest', value: '1', id: 'vote-slight-interest' },
  { label: 'Strong Interest', value: '2', id: 'vote-strong-interest' },
];
```

nguernse replied:
I like this suggestion! I've updated based on input.

helmturner commented on src/features/vote/proposals/list/ProposalInterestVote.tsx:
```diff
@@ -0,0 +1,95 @@
+        return (
+          <Label
+            key={`${proposal.id}-${option.id}`}
+            className={cn('flex items-center gap-2', {
```
POLISH: The radio buttons look pretty wonky to me. That's mostly because we have an odd number of off-balance items, but I wonder moving the labels below the radio buttons might help it look more even?
SUGGESTION:
```suggestion
            className={cn('flex w-[20%] flex-col text-center items-center gap-2', {
```

nguernse replied:
You're right, I agree with that. I've made an update to reflect this.

helmturner reviewed (changes requested):
PRAISE: excellent work again! Always a pleasure reviewing your code.
POLISH: I think this is g2g after just some minor touch-ups.
QUESTION: If you've got any opinions / curiosity about the form/fetching infra, I've got an [experimental branch](https://github.com/tulsawebdevs/website/tree/feat/form-experiments) where I'm exploring some options.

helmturner reviewed (approved):
PRAISE: I have nothing to add... this is just great. Beautiful work! Love the code-quality improvements.
